sjsonnet/src/sjsonnet/YamlRenderer.scala

```diff
@@ -0,0 +1,554 @@
+  var topLevel = true
+  var leftHandPrefixOffset = 0
+  var firstElementInArray = false
+  val newlinePattern = Pattern.compile("\n")
+
+  override def visitString(s: CharSequence, index: Int) = {
+    addSpaceAfterColon()
+    flushBuffer()
+
+    // Although we can render strings in a multitude of different ways, here
+    // we try our best to match PyYAML's style. Their style generally looks
+    // pretty reasonable, and allows consistency with other tooling a significant
+    // fraction of which is probably Python
+    val str = s.toString
+
+    // empty strings and single-newline strings are special-cased
+    if (str.isEmpty) out.append("''")
+    else if (str == "\n") out.append("|2+\n")
+    // Strings with trailing spaces or with unicode characters are written double-quoted
+    else if (str.contains(" \n") || str.exists(_ > '~'))
+      PrettyYamlRenderer.writeDoubleQuoted(out, indent * (depth + 1), leftHandPrefixOffset, idealWidth, str)
+    // Other strings with newlines are rendered as blocks
+    else if (str.contains('\n'))
+      PrettyYamlRenderer.writeBlockString(str, out, depth, indent)
+    // Strings which look like booleans/nulls/numbers/dates/etc.,
+    // or have leading/trailing spaces, are rendered single-quoted
+    else if (PrettyYamlRenderer.stringNeedsToBeQuoted(str)) {
+      val strWriter = new StringWriter
+      ujson.Renderer.escape(strWriter, s, unicode = true)
+      val quotedStr = "'" + str.replace("'", "''") + "'"
+      PrettyYamlRenderer.writeWrappedString(quotedStr, leftHandPrefixOffset, out, indent * (depth + 1), idealWidth)
+      leftHandPrefixOffset = quotedStr.length + 2
+    } else { // All other strings can be rendered naked without quotes
+      PrettyYamlRenderer.writeWrappedString(str, leftHandPrefixOffset, out, indent * (depth + 1), idealWidth)
+      leftHandPrefixOffset = s.length
+    }
+    out
+  }
+
+  def addSpaceAfterColon() = {
+    if (afterColon) {
+      out.append(' ')
+      afterColon = false
+    }
+  }
+  override def visitFloat64(d: Double, index: Int) = {
+    addSpaceAfterColon()
+    flushBuffer()
+    out.append(RenderUtils.renderDouble(d))
+    out
+  }
+
+  override def visitTrue(index: Int) = {
+    addSpaceAfterColon()
+    super.visitTrue(index)
+  }
+
+  override def visitFalse(index: Int) = {
+    addSpaceAfterColon()
+    super.visitFalse(index)
+  }
+
+  override def visitNull(index: Int) = {
+    addSpaceAfterColon()
+    super.visitNull(index)
+  }
+  override def flushBuffer() = {
+    if (newlineBuffered) {
+      afterColon = false
+      PrettyYamlRenderer.writeIndentation(out, indent * depth)
+    }
+    if (dashBuffered) {
+      out.append("- ")
+    }
+    dashBuffered = false
+    newlineBuffered = false
+    dashBuffered = false
+  }
+  override def visitArray(length: Int, index: Int) = new ArrVisitor[Writer, Writer] {
+    var empty = true
+    val dedentInObject = afterKey && !indentArrayInObject
+    def subVisitor = {
+      if (empty){
```

Comment:
In both `visitArray` and `visitObject`, we move some of the logic into when `subVisitor` is called the first time in order to avoid it being run if the array or object is empty. This is because empty arrays and objects are rendered as inline `[]`s or `{}`s, pretty different from the multi-line `key:`-delimited object or `- `-delimited list syntax, but due to the nature of the visitor we do not know whether an array or object is empty until someone calls `visitEnd`